### PR TITLE
gh-140287: Handle `PYTHONSTARTUP` script exceptions in the asyncio REPL

### PR DESCRIPTION
**Pull request by johnslavik**
~~This is generally a work in progress; tests are needed.~~
~~The asyncio patch is simple and ready.~~

* Issue: gh-140287

### CODE REVIEW

**Comment by bswck** [comment]
This could be even fancier and reuse `console.runcode`, technically.
Or... maybe not. I don't think we are supporting `PYTHONSTARTUP` scripts with top-level awaits.

**Comment by johnslavik** [comment]
Please merge #140298 first.

File: Lib/test/test_repl.py [hunk 1]
```diff
@@ -42,7 +43,11 @@ def spawn_repl(*args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, custom=F
     # path may be used by PyConfig_Get("module_search_paths") to build the
     # default module search path.
     stdin_fname = os.path.join(os.path.dirname(sys.executable), "<stdin>")
-    cmd_line = [stdin_fname, '-I']
+    cmd_line = [stdin_fname]
+    # Isolated mode implies -E, -P and -s, purifies sys.path and ignores PYTHON*
+    # variables.
+    if isolated:
+        cmd_line.append('-I')
```
Comment (johnslavik) [inline Lib/test/test_repl.py:49]: To reuse this routine and `spawn_asyncio_repl` below, we need to have a knob to not pass `-I` since that ignores `PYTHONSTARTUP` completely.

The asyncio REPL currently doesn't comply to that, which is a separate issue I'll report having learnt this.

File: Lib/test/test_repl.py [hunk 2]
```diff
@@ -292,24 +235,6 @@ def bar(x):
...
-            env = os.environ.copy()
-            env["PYTHON_HISTORY"] = os.path.join(tmpdir, ".asyncio_history")
-            env["PYTHONSTARTUP"] = script
-            subprocess.check_call(
-                [sys.executable, "-m", "asyncio"],
-                stdout=subprocess.PIPE,
-                stderr=subprocess.PIPE,
-                env=env,
-                timeout=SHORT_TIMEOUT,
-            )
```
Comment (johnslavik) [inline Lib/test/test_repl.py]: Covered by `TestPythonStartup` now, so not needed here.

File: Lib/test/test_repl.py (hunk 1 above)
Comment (johnslavik) [reply Lib/test/test_repl.py:49]: Tracking the `-I` edge case in #140648.

File: Lib/test/test_repl.py [hunk 3]
```diff
@@ -346,6 +271,64 @@ def test_asyncio_repl_is_ok(self):
...
+                new_startup_env(code=startup_code, histfile=histfile) as startup_env
+            ):
+                p = repl_factory(env=os.environ | startup_env, isolated=False)
+                p.stdin.write('foo()')
+                output = kill_python(p)
+
+                for expected in (
+                    "Traceback (most recent call last):",
+                    'File "<stdin>", line 1, in <module>',
+                    f'File "{startup_env['PYTHONSTARTUP']}", line ',
```
Comment (johnslavik) [inline Lib/test/test_repl.py]: I need to understand why line number 2 fails specifically on Windows and 3 is expected.

**Review by johnslavik — COMMENTED** [review]
I'm also thinking that ideally we would test stdout and stderr separately.

File: Lib/test/test_repl.py (hunk 3 above)
Comment (johnslavik) [reply Lib/test/test_repl.py]: Still not sure. Will see

**Comment by johnslavik** [comment]
This is likely it. I can't think of anything simpler but equally complete at the same time.
This suite seems intentionally straightforward, as are the new tests proposed here.
I'll give this a few more days in a daemon thread of my mind.

We can abstract as needed in the future. We could use some regexes.
However, I don't think this is so much needed now, and the environments differ enough to justify some repetition.

File: Lib/test/test_repl.py [hunk 4]
```diff
@@ -365,6 +341,7 @@ def f():
         self.assertEqual(traceback_lines, expected_lines)
 
 
+@support.force_not_colorized_test_class
```
Comment (bswck) [inline Lib/test/test_repl.py]: I'm not sure this is needed, but I don't think it's bad to keep it as is.

**Comment by bswck** [comment]
I'm seeing the same intriguing failures on Windows. Seems like newly discovered bugs:
1. Doesn't linecache store PYTHONSTARTUP lines on Windows?
2. Why are faulty lines in PYTHONSTARTUP tracebacks on Windows off-by-one?

**Comment by johnslavik** [comment]
It seems that the asyncio REPL on Windows leaks the event loop:

```
❯ PYTHONSTARTUP='' python -m asyncio
Running Debug|x64 interpreter...
asyncio REPL 3.15.0a1+ (heads/asyncio-repl-handle-python-startup-dirty:ce03ccef689, Nov  8 2025, 02:06:1) [MSC v.1944 64 bit (AMD64)] on win32
Use "await" directly instead of "asyncio.run()".
Type "help", "copyright", "credits" or "license" for more information.
>>> import asyncio
>>> exit
exiting asyncio REPL...
C:\Users\zenon\Python\cpython\Lib\asyncio\base_events.py:758: ResourceWarning: unclosed event loop <ProactorEventLoop running=False closed=False debug=False>
```

Definitely a bug, we'll track it elsewhere soon.

EDIT: Tracking in #141262

**Comment by bswck** [comment]
> Doesn't linecache store PYTHONSTARTUP lines on Windows?

It seems that the reason why it doesn't show up is because it's in temp dir.

**Comment by bswck** [comment]
I'm not sure why on Windows removing the temporary file before displaying the traceback makes the lines unavailable, while on other platforms it doesn't matter.

**Comment by bswck** [comment]
I'll start with a small reproducer for Linux and Windows regarding file deletion and linecache.
When it comes to the off-by-one, I'm still investigating.

**Comment by johnslavik** [comment]
This is ready, but Windows doesn't pass because of GH-141262.

**Comment by johnslavik** [comment]
It seems that these unclosed loop warnings regressed (stopped appearing) in the recent releases -- this is bad, but we can fix them separately and we can also fix the asyncio REPL in this regard by actually closing the loop (in a separate PR).

In the meantime, updating this branch should make the CI green.

**Comment by johnslavik** [comment]
Looks like I was confused about the regression -- that's good news!

This will have to wait until the loop is properly closed in the asyncio REPL.

**Comment by johnslavik** [comment]
Cherry-picked 92f629b (875fd2a) from GH-142785 to see if it fixes the problem.

**Comment by johnslavik** [comment]
> to see if it fixes the problem.

It does! Cool!

File: Lib/asyncio/__main__.py [hunk 5]
```diff
@@ -101,7 +101,12 @@ def run(self):
                 import tokenize
                 with tokenize.open(startup_path) as f:
                     startup_code = compile(f.read(), startup_path, "exec")
-                    exec(startup_code, console.locals)
+                    try:
+                        exec(startup_code, console.locals)
+                    except SystemExit:
+                        raise
```
Comment (johnslavik) [inline Lib/asyncio/__main__.py]: I *really* don't like this, but it resembles the original behavior of the REPLs -- see GH-143023.
I don't think this is correct, but it's not super clear to me, so please chime in to GH-143023 to put your two cents in.

```suggestion
                    # TODO: Revisit in GH-143023
                    except SystemExit:
                        raise
```

Comment (johnslavik) [reply Lib/asyncio/__main__.py]: GH-143023 is also why I didn't add a test for this.

**Comment by johnslavik** [comment]
I'll see if I can simplify this. I think the tests are slightly overcomplicated.

**Review by johnslavik — COMMENTED** [review]
I'll come back to this after completing GH-140648.

File: Lib/test/test_repl.py (hunk 1 above)
Comment (johnslavik) [reply Lib/test/test_repl.py:49]: No longer a part of this PR.

File: Lib/asyncio/__main__.py (hunk 5 above)
Comment (johnslavik) [reply Lib/asyncio/__main__.py]: We've got some good news, this is not bad! I'll add a test shortly.

**Comment by johnslavik** [comment]
I've realized that this isn't testing what it is supposed to test. I'll come back to this later.

File: Lib/test/test_repl.py [hunk 6]
```diff
@@ -67,6 +68,19 @@ def spawn_repl(*args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, custom=F
...
+def new_pythonstartup_env(*, code: str, histfile: str = ".pythonhist"):
+    """Create environment variables for a PYTHONSTARTUP script in a temporary directory."""
+    with os_helper.temp_dir() as tmpdir:
+        filename = os.path.join(tmpdir, "pythonstartup.py")
+        with open(filename, "w") as f:
+            f.write(code)
+        yield {
+            "PYTHONSTARTUP": filename,
+            "PYTHON_HISTORY": os.path.join(tmpdir, histfile)
+        }
```
Comment (johnslavik) [inline Lib/test/test_repl.py:81]: This will be useful in https://github.com/python/cpython/issues/143023. I can inline it in `test_pythonstartup_failure` until then if we want to increase locality and not overdo idioms prematurely.

File: Lib/asyncio/__main__.py (hunk 5 above)
Comment (johnslavik) [reply Lib/asyncio/__main__.py]: I'll not be adding tests for this; the exact handling of exception types in repl startup is still up for debate.

File: Lib/test/test_repl.py [hunk 7]
```diff
@@ -449,6 +461,32 @@ def test_quiet_mode(self):
...
+        startup_env = self.enterContext(
+            temp_pythonstartup(source=startup_code, histfile=".asyncio_history"))
+
+        p = spawn_repl(
+            "-qm", "asyncio",
+            env=os.environ | startup_env,
+            isolated=False,
+            custom=True)
+        p.stdin.write("print('user code', 'executed')\n")
+        output = kill_python(p)
```
Comment (pablogsal) [inline Lib/test/test_repl.py:482]: This new test verifies the traceback text and that user code ran afterward, but never asserts `p.returncode == 0`. The whole point of this PR is that a failing `PYTHONSTARTUP` script is handled gracefully and the REPL continues normally. Every comparable asyncio REPL test in this file asserts the return code but this one is the odd one out.

File: Lib/asyncio/__main__.py [hunk 8]
```diff
@@ -101,11 +101,16 @@ def run(self):
...
-
-                import tokenize
-                with tokenize.open(startup_path) as f:
-                    startup_code = compile(f.read(), startup_path, "exec")
+                try:
+                    import tokenize
+                    with tokenize.open(startup_path) as f:
+                        startup_code = compile(f.read(), startup_path, "exec")
                     exec(startup_code, console.locals)
+                # TODO: Revisit in GH-143023
```
Comment (pablogsal) [inline Lib/asyncio/__main__.py]: https://github.com/python/cpython/issues/143023 is closed as not planned. What you plan t do here?

File: Lib/test/test_repl.py (hunk 7 above)
Comment (johnslavik) [reply Lib/test/test_repl.py:482]: LOL I literally wrote this class and missed it. Fixed!

File: Lib/asyncio/__main__.py (hunk 8 above)
Comment (johnslavik) [reply Lib/asyncio/__main__.py]: Forgot to clean this up, thx
